File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void reverseList(ListNode* s,ListNode* e){
        ListNode *c = s,*p=NULL,*n=NULL;
        while(c!=e){
            n = c->next;
            c->next = p;
            p = c;
            c = n;
        }
        c->next = p;
    }
    ListNode* reverseKGroup(ListNode* head, int k) {
     if(head==nullptr || head->next==nullptr || k==1) return head;
        ListNode *s = head, *e = head;
        int i = k-1;
        while(i!=0){
            e=e->next;
            if(e==nullptr) return head;
            i--;
        }
        ListNode *rest = reverseKGroup(e->next,k);
        reverseList(s,e);
        s->next = rest;
        return e;
    }
};

File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reverseList(ListNode start,ListNode end){
        ListNode curr = start;
        ListNode prev = null;
        ListNode nex  = null;
        while(curr!=end){
            nex = curr.next;
            curr.next = prev;
            prev = curr;
            curr = nex;
        }
        curr.next=prev;
    }
    public ListNode reverseKGroup(ListNode head, int k) {
        if(head==null || head.next==null || k==1) return head;
        ListNode dummyHead = new ListNode(0,head);
        ListNode beforeStart = dummyHead;
        ListNode end = head;
        int i=0;
        while(end!=null){
            i++;
            if(i%k==0){
                ListNode start = beforeStart.next;
                ListNode afterEnd = end.next;
                reverseList(start,end);
                beforeStart.next = end;
                start.next = afterEnd;
                beforeStart = start;
                end = afterEnd;
            }
            else
                end = end.next;
        }
        return dummyHead.next;
    }
}